Adicionando 1868 (#183)
petbccufscar/URI: Strings/1868/README.md

![imagemexemplo](https://resources.urionlinejudge.com.br/gallery/images/problems/UOJ_1868.png)

Escreva um programa que, dado um número inteiro, imprima todos os quadros da animação da espiral quadrada.

**Problema completo**: https://www.urionlinejudge.com.br/judge/pt/problems/view/1868 

# Resolução:

Para resolver esse problema, a lógica é a seguinte: de ínicio deve-se criar, conforme as especificações do teste em questão, uma matriz, completá-la com 'O' e em seu centro colocar um 'X'. Em seguida, a partir da atual posição de 'X', define-se bordas de movimentação do caractere, tais bordas ditarão os limites de movimento - para a direita, cima, esquerda e baixo - do caractere dentro da matriz. Depois, expande-se gradativamente as bordas até que o caractere tenha percorrido todos os elementos da matriz. Dessa forma, a movimentação assume um formato de espiral.  

> Antes de mais nada, declara-se as bibliotecas de função que serão usadas. Neste caso, apenas a biblioteca `<stdio.h>`, para receber dados e imprimir informações na tela.
> ``` c
> #include <stdio.h>
> ```

Reviewer: ```suggestion
```
Não precisa dizer q esta usando essa biblioteca pq ela é padrão. Se quiser encurtar um poquinho kkkk

Como mostrou-se necessário mover o caractere 'X', é aconselhável criar um procedimento, do tipo `void`, para completar a matriz com 'O', pois assim a cada movimento de caractere chama-se o procedimento para reiniciar a matriz(colocar 'O' em todos seus elementos). Dessa forma, não é preciso fazer trocas entre caractere 'X' e 'O', apenas posicionar 'X' em seu devido lugar. 
Para o procedimento em si, nomeia-se como `preenche_matriz`. Em seu interior, declara-se duas variáveis, para loops, `i` e `j`. Depois, abri-se um loop `for` para acessar cada linha da matriz, dentro dele, aninha-se outro `for` para acessar cada coluna da matriz. Dessa forma, com uso de `i` e `j`, acessa-se cada elemento da matriz e atribui-se a eles o caractere 'O'.
``` c
void preenche_matriz (char matriz[25][25], int casos_teste) {
    int i, j;
    for (i=0; i<casos_teste; i++) {
            for(j=0; j<casos_teste; j++) {
                matriz[i][j] = 'O';
            }
        }
}
```

> Agora, para cada movimento de 'X', inclusive seu posicionamento central inicial, deve-se imprimir a matriz completamente. Para tal, devido a repetitiva necessidade de impressão, é aconselhável que se faça um procedimento, do tipo `void`, para imprimir a matriz por completo, pois assim basta chamá-lo a cada movimentação de caracte. Dessa forma, nomeia-se o procedimento como `imprime_matriz`. Nele, as declarações e a lógica para se acessar cada elemento do procedimento anterior(`preenche_matriz`) se repete. Contudo, ao invés de atribuir se valor a cada elemento, imprimi-se, com uso de `scanf`, o elemento. Ademais, ao final do acesso de todos os elementos de cada linha, imprimi-se uma quebra de linha e, ao final da impressão de todos os elementos, imprimi-se um '@' com quebra de linha - finalizando as impressões.

Reviewer: ```suggestion
Agora, para cada movimento de 'X', inclusive seu posicionamento central inicial, deve-se imprimir a matriz completamente. Para tal, devido a repetitiva necessidade de impressão, é aconselhável que se faça um procedimento, do tipo `void`, para imprimir a matriz por completo, pois assim basta chamá-lo a cada movimentação de caracte. Dessa forma, nomeia-se o procedimento como `imprime_matriz`. Nele, as declarações e a lógica para se acessar cada elemento é igual ao procedimento anterior (`preenche_matriz`). Contudo, ao invés de atribuir se valor a cada elemento, imprimi-se, com uso de `printf`, o elemento. Ademais, ao final do acesso de todos os elementos de cada linha, imprimi-se uma quebra de linha e, ao final da impressão de todos os elementos, imprimi-se um '@' com quebra de linha - finalizando as impressões.
```

``` c
int main () {
    int casos_teste, percorre, direita_borda, esquerda_borda, superior_borda, inferior_borda, linha, coluna;
    char matriz[25][25];
    /* resto do código omitido para futura explicação */
```

Na sequência, lê-se o primero caso de teste, que também é a definição da dimensão da matriz a ser usada. Depois, abri-se um loop `while` que executará enquanto a entrada para o caso de teste for diferente de 0. Iniciando o loop, chama-se o procedimento para completar a matriz com caractere 'O'(daqui para frente chama-se tal procedimento de reiniciação da matriz). Após tal tarefa, defini-se o centro da matriz, para tal realiza-se a divisão exata da dimensão da matriz por 2, atribuindo tal valor para `percorre`. Dessa forma, defini-se a localização, nas linhas e colunas, onde 'X' deve ser colocado, pois `percorre` representa o centro delas. Portanto, atribui-se o caracter para o elemento central da matriz. Por fim, imprimi-se a matriz.
``` c
    scanf("%d", &casos_teste);
    while(casos_teste!= 0) {
        preenche_matriz(matriz, casos_teste);
        percorre = (casos_teste)/2;
        matriz[percorre][percorre] = 'X';
        imprime_matriz(matriz, casos_teste);
```

> Para continuação do problema, torna-se necessário definir as bordas de movimetação. Para tal, é importante ter em mente o funcionamento dos indíces de uma matriz na linguagem C. Sendo assim, para configurar a borda direita, tenha-se em mente que uma movimentação para a direita na matriz, na linguagem C, significa um incremento em relação ao valor de coluna, logo a borda recebe o valor da coluna central mais 1. Para definir a borda superior, tenha-se em mente que uma movimentação para cima significa um decremento em relação ao valor de linha, logo a borda recebe o valor de linha central menos 1. Para configurar a borda esquerda, tenha-se em mente que uma movimentação para a esquerda significa um decremento em relação ao valor das colunas, logo a borda recebe o valor de coluna central menos 1. Para definir a borda inferior, tenha-se em mente que uma movimentação para baixo significa um incremento em relação ao valor de linha, logo a borda recebe o valor de linha central mais 1. Ademais, para futuro controle, atribui-se às variáveis `linha` e `coluna` o valor de posição, da linha e coluna respectivamente, atual do caractere 'X', nesse caso o valor de percorre.

Reviewer: ```suggestion
Para continuação do problema, torna-se necessário definir as bordas de movimetação. Para tal, é importante ter em mente o funcionamento dos indíces de uma matriz na linguagem C. Sendo assim, para configurar a borda direita, na linguagem C, significa um incremento em relação ao valor de coluna, logo a borda recebe o valor da coluna central mais 1. Para definir a borda superior, tenha-se em mente que uma movimentação para cima significa um decremento em relação ao valor de linha, logo a borda recebe o valor de linha central menos 1. Para configurar a borda esquerda, uma movimentação para a esquerda significa um decremento em relação ao valor das colunas, logo a borda recebe o valor de coluna central menos 1. Para definir a borda inferior, é necessário uma movimentação para baixo, o que significa um incremento em relação ao valor de linha, logo a borda recebe o valor de linha central mais 1. Ademais, para futuro controle, atribui-se às variáveis `linha` e `coluna` o valor de posição, da linha e coluna respectivamente, atual do caractere 'X', nesse caso o valor de percorre.
```

``` c
        direita_borda = percorre + 1;
        superior_borda = percorre - 1;
        esquerda_borda = percorre - 1;
        inferior_borda = percorre + 1;
        linha = percorre;
        coluna = percorre;
``` 

> Com as bordas definidas, parte-se para a questão da movimentação em si. Como 'X' irá se movimentar diversas vezes, abri-se um loop `while` para que a movimentação seja automatizada e aconteça até que a borda direita limitante seja igual à dimensão da matriz. Ademais, a borda limitante deve ser a direita, pois a última impressão deve ter o caractere 'X' no elemento mais inferior e mais a direita da matriz. 

Reviewer: ```suggestion
Com as bordas definidas, parte-se para a questão da movimentação em si. Como 'X' irá se movimentar diversas vezes, abri-se um loop `while` para que a movimentação seja automatizada e aconteça até que a borda direita limitante seja igual à dimensão da matriz. Ademais, a borda limitante deve ser a direita, pois a última impressão deve ter o caractere 'X' como o elemento mais inferior e mais a direita da matriz. 
```

``` c
        while (direita_borda < casos_teste) {
            /* resto do código omitido para futura 
            explicação */
```

> Dando prosseguimento, para cada passo de 'X' - direita, para cima, esquerda, para baixo - cria-se um loop `while` para controle e automação. Logo, para movimentos a direita a delimitação será a borda direita já estabelecida. Dentro do loop, reinicia-se a matriz e incrementa-se `percorre` para que possa haver movimentação à direita. Assim, com o uso da informação armazenada em `linha` e o atual `percorre`, posiciona-se o 'X' no seu novo elemento e imprimi-se a matriz. Dessa forma, o loop realiza os movimentos até que a borda seja atingida, configurando o movimento espiral. Ao final do loop se torna necessário atualizar as informações sobre a coluna, logo a variável de controle recebe o valor da borda direita, incrementar o valor da borda direita, para a próxima movimentação, e configurar o valor de `percorre` para a posição da atual linha - possibilitando o próximo movimento que usa o indice da linha.

Reviewer: ```suggestion
Dando prosseguimento, para cada passo de 'X' - direita, para cima, esquerda, para baixo - cria-se um loop `while` para controle e automação. Logo, para movimentos a direita a delimitação será a borda direita já estabelecida. Dentro do loop, reinicia-se a matriz e incrementa-se `percorre` para que possa haver movimentação à direita. Assim, com o uso da informação armazenada em `linha` e o atual `percorre`, posiciona-se o 'X' em sua nova posição e imprimi-se a matriz. Dessa forma, o loop realiza os movimentos até que a borda seja atingida, configurando o movimento espiral. Ao final do loop se torna necessário atualizar as informações sobre a coluna, logo a variável de controle recebe o valor da borda direita, o valor da borda direita é incrementado, para a próxima movimentação, e o valor de `percorre` recebe a posição da atual linha - possibilitando o próximo movimento que usa o indice da linha.
```

``` c
            while(percorre > superior_borda) {
                preenche_matriz(matriz, casos_teste);
                percorre--;
                matriz[percorre][coluna] = 'X';
                imprime_matriz(matriz, casos_teste);
            }
            coluna=direita_borda;
            direita_borda++;
            percorre = linha;  
```

> Essa lógica utilizada se repete para cada movimentação, apenas com pequenas adaptações, quando necessário, de código para respeitar a questão dos índices de acesso aos elementos da matriz. Ou seja, após a movimentos para baixo e esquerda, a borda superior e esquerda devem diminuir; aṕos movimentos para cima, a borda superior deve aumentar. E para respeitar a execução dos movimentos seguintes. Ou seja, quando os movimentos são para a esquerda e direita, salva-se a posição da coluna em que se terminou e a variável de posição recebe a linha; quando os movimentos são para cima ou para baixo, salva-se a posição da linha em que se terminou e a variável de posição recebe a coluna. Assim sendo, termina-se o loop da espiral.

Reviewer: ```suggestion
Essa lógica utilizada se repete para cada movimentação, apenas com pequenas adaptações, quando necessário, de código para respeitar a questão dos índices de acesso aos elementos da matriz. Ou seja, após os movimentos para baixo e esquerda, as bordas superior e esquerda devem diminuir; aṕos movimentos para cima, a borda superior deve aumentar. Para respeitar a execução dos movimentos seguintes, ou seja, quando os movimentos são para a esquerda e direita, são salvas a posição da coluna em que se terminou e a variável de posição recebe a linha; quando os movimentos são para cima ou para baixo, são salvas a posição da linha em que se terminou e a variável de posição recebe a coluna. Assim sendo, termina-se o loop da espiral.
```

``` c
            while(percorre > superior_borda) {
                preenche_matriz(matriz, casos_teste);
                percorre--;
                matriz[percorre][coluna] = 'X';
                imprime_matriz(matriz, casos_teste);
            }
            linha = superior_borda;
            superior_borda--;
            percorre = coluna;
            while(percorre > esquerda_borda) {
                preenche_matriz(matriz, casos_teste);
                percorre--;
                matriz[linha][percorre] = 'X';
                imprime_matriz(matriz, casos_teste);
            }
            coluna = esquerda_borda;
            esquerda_borda--;
            percorre=linha;
            while(percorre < inferior_borda) {
                preenche_matriz(matriz, casos_teste);
                percorre++;
                matriz[percorre][coluna] = 'X';
                imprime_matriz(matriz, casos_teste);
            }
            linha = inferior_borda;
            inferior_borda++;
            percorre = coluna;         
        }
``` 

> Apesar de terminado o loop, ainda resta uma última movimentação para a direita, porque o movimento espiral inicia e termina para direita. Tal processo deve ser feito de forma única e fora do loop, pois assim não se inicia movimentações desnecessárias para os outros sentidos. A lógica e código é são identicos às utilizadas no interior do loop. Por fim, lê-se novamente o valor de `casos_teste` para se reiniciar ou terminar os teste, e, por consequência o loop principal. Ademais, se retorna 0 da função `main` para finalização do programa.

Reviewer: ```suggestion
Apesar de terminado o loop, ainda resta uma última movimentação para a direita, porque o movimento espiral inicia e termina para direita. Tal processo deve ser feito de forma única e fora do loop, pois assim não se inicia movimentações desnecessárias para os outros sentidos. A lógica e código são identicos às utilizadas no interior do loop. Por fim, lê-se novamente o valor de `casos_teste` para se reiniciar ou terminar os testes, e, por consequência o loop principal. Ademais, se retorna 0 da função `main` para finalização do programa.
```
